pkg/handlers/pdf.go

```diff
@@ -2,96 +2,54 @@ package handlers
 
 import (
 	"bytes"
+	"context"
 	"encoding/json"
-	"fmt"
 	"io"
 	"io/ioutil"
-	"mime/multipart"
 	"net/http"
-	"net/textproto"
-	"os"
 )
 
+type invokeLambdaFunc func(cxt context.Context, html string) ([]byte, error)
+
 // PDFHandler handles PDFs
 type PDFHandler struct {
 	HandlerBase
+	invokeLambda invokeLambdaFunc
 }
 
 // NewPDFHandler returns a new PDFHandler
-func NewPDFHandler() *PDFHandler {
-	return &PDFHandler{}
+func NewPDFHandler(invoker invokeLambdaFunc) *PDFHandler {
+	return &PDFHandler{invokeLambda: invoker}
+}
+
+type requestPayload struct {
+	HTML string `json:"html"`
 }
 
 // Handle returns an http.HandlerFunc
 func (h PDFHandler) Handle() http.HandlerFunc {
 	return func(w http.ResponseWriter, r *http.Request) {
-		type GenerateRequest struct {
-			HTML     string `json:"html"`
-			Filename string `json:"filename"`
-		}
-
-		var data GenerateRequest
 
 		body, readErr := ioutil.ReadAll(r.Body)
 		if readErr != nil {
 			h.WriteErrorResponse(r.Context(), w, readErr)
 			return
 		}
 
-		parseErr := json.Unmarshal(body, &data)
+		var generateRequest requestPayload
+		parseErr := json.Unmarshal(body, &generateRequest)
 		if parseErr != nil {
 			h.WriteErrorResponse(r.Context(), w, parseErr)
 			return
 		}
 
-		postBody := new(bytes.Buffer)
-		writer := multipart.NewWriter(postBody)
-
-		header := make(textproto.MIMEHeader)
-		header.Set("Content-Disposition", `form-data; name="file"; filename="input.html"`)
-		header.Set("Content-Type", "text/html")
-
-		part, partErr := writer.CreatePart(header)
-		if partErr != nil {
-			h.WriteErrorResponse(r.Context(), w, partErr)
-			return
-		}
-		if _, partWriteErr := part.Write([]byte(data.HTML)); partWriteErr != nil {
-			h.WriteErrorResponse(r.Context(), w, partWriteErr)
+		result, generateErr := h.invokeLambda(r.Context(), generateRequest.HTML)
```

Comment:
I'm assuming you've already thought through this, but I just feel like I have to ask. Is there a security concern here regarding receiving, processing, and operating on arbitrary HTML?

Reply:
Part of running this in a lambda is that we can lock everything down and prevent the lambda from having access to anything. We're also running Prince [with `--no-network` to prevent it from making any outbound network connections](https://github.com/CMSgov/easi-infra-modules/blob/master/lambda/prince/handler.go#L62).

One downside to Prince being closed source is that we can't audit its parsers, etc. So we will isolate it as much as possible.